src/pip/_internal/utils/parallel.py

```diff
@@ -0,0 +1,215 @@
+"""Convenient parallelization of higher order functions.
+
+This module provides proper fallback functions for multiprocess
+and multithread map, both the non-lazy, ordered variant
+and the lazy, unordered variant.
+"""
+
+__all__ = ['map_multiprocess', 'imap_multiprocess',
+           'map_multithread', 'imap_multithread']
+
+from contextlib import contextmanager
+from multiprocessing import Pool as ProcessPool
+from multiprocessing.dummy import Pool as ThreadPool
+
+from pip._vendor.requests.adapters import DEFAULT_POOLSIZE
+from pip._vendor.six import PY2
+from pip._vendor.six.moves import map
+
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import (
+        Callable, Iterable, Iterator, List, Optional, Union, TypeVar)
+    from multiprocessing import pool
+
+    Pool = Union[pool.Pool, pool.ThreadPool]
+    S = TypeVar('S')
+    T = TypeVar('T')
+
+# On platforms without sem_open, multiprocessing[.dummy] Pool
+# cannot be created.
+try:
+    import multiprocessing.synchronize  # noqa
+except ImportError:
+    LACK_SEM_OPEN = True
+else:
+    LACK_SEM_OPEN = False
+
+# Incredibly large timeout to work around bpo-8296 on Python 2.
+TIMEOUT = 2000000
+
+
+@contextmanager
+def closing(pool):
```

Comment:
This isn't necessary, since pool's __exit__ should handle closing and other details.

Reply:
It's needed for `imap*` to start submitting tasks to the pool for some reason.  The non lazy variant doesn't need it though.  I think I'll add a comment explaining why it is needed.